text/0000-pack-subcommands.md

```diff
@@ -0,0 +1,186 @@
+  trust-builder         Trust builder
+  untrust-builder       Stop trusting builder
+  list-trusted-builders List Trusted Builders
+  create-builder        Create builder image
+  package-buildpack     Package buildpack in OCI format.
+  suggest-stacks        Display list of recommended stacks
+  version               Show current 'pack' version
+  report                Display useful information for reporting an issue
+  completion            Outputs completion script location
+  help                  Help about any command
+```
+
+# What it is
+[what-it-is]: #what-it-is
+
+
+### Proposed structure
+
+```
+pack
+
+  app
+    build
+    inspect
+    rebase
+  builder
+    create
+    inspect
+    suggest
+  buildpack
+    package
+    register
+  completion
+  config
+    default-builder
+      set
+      unset
+      show
+    run-image-mirror
+      add
+      remove
+      list
+    registry
+      add
+      remove
+      list
+    trusted-builder
+      add
+      remove
+      list
+  help
+  report
+  stack
+    suggest
+  version
+```
+
+### Top-Level Aliases
+
+There would also be some alias subcommands for the `app` set of subcommands for easier usage and compatibility.
+
+* `pack build` -> `pack app build`
+* `pack inspect` -> `pack app inspect`
+* `pack rebase` -> `pack app rebase`
+
+### Config
+
+The `config` subcommand would work with the notion of two different type of resources:
+
+##### Singular
+
+Singular resources would have the following subcommands:
+
+* set
+* unset
+* show (aliased to subcommand root)
+
+###### Example
+
+```
+# sets the default builder
+pack config default-builder set gcr.io/my-org/my-builder:latest
+
+# unset the default builder
+pack config default-builder unset
+
+# show the default builder
+pack config default-builder show
+
+# show the default builder (w/o subcommand)
+pack config default-builder
+```
+
+##### List
+
+List resources would have the following subcommands:
+
+* add
+* remove
+* list
+
+The explicit additional parameters would be based on the specific implementation needs.
+
+###### Example
+
+```
+# adds `gcr.io/my-org/my-image` as a mirror for `my-image`  
+pack config run-image-mirrors add my-image gcr.io/my-org/my-image
+
+# removes `gcr.io/my-org/my-image` as a mirror for `my-image`
+pack config run-image-mirrors remove my-image gcr.io/my-org/my-image
+
+# removes all mirrors for `my-image`
+pack config run-image-mirrors remove my-image
+
+# list all mirrors
+pack config run-image-mirrors list
+
+# list all mirrors for `my-image`
+pack config run-image-mirrors list my-image
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+
+## Backwards Compatibility
+
+1. There would be a period in which all pre-existing subcommands would work.
+2. [Top-level aliases](#top-level-aliases) will provide continued compatibility for common operations.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Change is bad... mkay.
+
+# Alternatives
+[alternatives]: #alternatives
+
```

Comment:
Another alternative is to focus "pack" more for specific audiences which would reduce the number of commands that it would have to know.

Reply:
What would that look like?

Assuming `pack` remains the platform I could foresee the following commands:

```
pack

  build
  config
     ...
  inspect
  run
```

The alternative utility CLI would be everything else:

```
bputils

  builder
    create
    inspect
    suggest
  buildpack
    create
    package
    register
  stack
    suggest
```

Reply:
```suggestion

### Alternative Utility

> Another alternative is to focus "pack" more for specific audiences which would reduce the number of commands that it would have to know.

<code>
pack

  build
  config
     ...
  inspect
  run
</code>

The alternative utility CLI would be everything else:

<code>
bputils

  builder
    create
    inspect
    suggest
  buildpack
    create
    package
    register
  stack
    suggest
</code>

```